### ISSUE TITLE
[BUG] OOM occurs when ServiceBusSessionReceiverClient is repeatedly created and closed for Azure service bus

### ISSUE THREAD

**Issue by zafar142007** [Azure/azure-sdk-for-java#23539]
**Describe the bug**
When `ServiceBusSessionReceiverClient`s and from them `ServiceBusReceiverClient`s are repeatedly created to read sessions and closed, after some time OOM occurs as `bounded-elastic-evictor` threads are created infinitely. 
Ideally the closing off of receivers should have closed off their respective threads, but in this case this is not happening. The close() methods of the above two classes should be fixed.

***Exception or Stack Trace***
OOM occurred 

**To Reproduce**
In a receiver class
1. Create a `ServiceBusClientBuilder`
2. In an infinite while loop, create a `ServiceBusSessionReceiverClient` and then accept a session to get a `ServiceBusReceiverClient`.
3. Close the  `ServiceBusSessionReceiverClient` and `ServiceBusReceiverClient` inside the while loop.
4. Run the above class. Connect JVisualVM to this process to see the number of live threads in the application. 
5. After a while you will see the threads increasing indefinitely.

Use the code snippet below.

***Code Snippet***

```package com.trial.asb;
import com.azure.core.amqp.AmqpRetryOptions;
import com.azure.core.amqp.AmqpTransportType;
import com.azure.messaging.servicebus.ServiceBusClientBuilder;
import com.azure.messaging.servicebus.ServiceBusReceiverClient;
import com.azure.messaging.servicebus.ServiceBusSessionReceiverClient;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import java.time.Duration;

public class BugReproduction {

    final static String connectionString = "/*ENTER CONNECTION STRING*/";
    final static String queue = "/*QUEUE_NAME*/";
    static Logger logger = LogManager.getLogger(BugReproduction.class);

    public static void main(String ar[]) {
        ServiceBusReceiverClient receiver = null;
        ServiceBusSessionReceiverClient sessionReceiverClient = null;
        ServiceBusClientBuilder sharedConnectionBuilder = new ServiceBusClientBuilder()
                .retryOptions(new AmqpRetryOptions().setTryTimeout(Duration.ofSeconds(10)).setMaxRetries(0))
                .transportType(AmqpTransportType.AMQP_WEB_SOCKETS)
                .connectionString(connectionString);
        while (!Thread.currentThread().isInterrupted()) {

            sessionReceiverClient = sharedConnectionBuilder
                    .sessionReceiver()
                    .queueName(queue)
                    .disableAutoComplete()
                    .buildClient();
            logger.info("Acquiring session");
            try {
                receiver = sessionReceiverClient.acceptNextSession();
            } catch (Exception e) {
                e.printStackTrace();
            } finally {
                // Dispose of our resources
                if (receiver != null) {
                    receiver.close();
                }
                if (sessionReceiverClient != null) {
                    sessionReceiverClient.close();
                }
                logger.info("Closed the receiver");
            }
        }
    }
}
```

**Expected behavior**
Closing off a `ServiceBusSessionReceiverClient` or a `ServiceBusReceiverClient` should not leak connections. It seems closing is not able to close off the underlying connection. These connections are increasing threads in the system. There should be a clean way to dispose off these client objects (including resources held by them).

**Screenshots**
<img width="1201" alt="Screenshot 2021-08-13 at 2 16 44 AM" src="https://user-images.githubusercontent.com/1270514/129267562-33714c69-f977-48d3-b005-af06cfa3175a.png">
<img width="1207" alt="Screenshot 2021-08-13 at 2 16 25 AM" src="https://user-images.githubusercontent.com/1270514/129267591-565b38b7-2954-4105-b17b-3573870257a8.png">

**Setup (please complete the following information):**
 - OS: [MacOS Catalina 10.15.7]
 - IDE: [Intellij]
 - Library/Libraries: [com.azure:azure-messaging-servicebus:7.3.0 (groupId:artifactId:version)]
 - Java version: [11] 
 - App Server/Environment: [Java main class running in Intellij]
 - Frameworks: [none] 

If you suspect a dependency version mismatch (e.g. you see `NoClassDefFoundError`, `NoSuchMethodError` or similar), please provide
- dependency tree (`mvn dependency:tree -Dverbose`)
NA
  
**Additional context**
The context is that we want to implement message consumption in strict ordering using Azure Service Bus's sessions. For messages belonging to a particular class we tag them with a sessionId and push them into a Service Bus queue. Then while reading a session we want the queue to be processed in the order of messages that were sent to the session. 

For this we saw that if we close off the clients the last abandoned message would be served as the first message the next time a client gets a lock onto this session. This is what prompted us to close the clients (`ServiceBusSessionReceiverClient` and `ServiceBusReceiverClient`) each time we read a session. But we ran into this OOM issue. 

The alternate way of reusing a client for reading off available sessions works but strict ordering is not maintained. In this case when we abandon a message in a session,  the next time a consumer gets a lock onto this session, the last abandoned message is most likely not the first message. So this way is unfeasible.

**Information Checklist**
Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [X] Bug Description Added
- [X] Repro Steps Added
- [X] Setup information Added

**Comment by joshfree** [comment]
Thanks for filing this github issue, @zafar142007.

@conniey this sounds similar to another issue that you're investigating.

/cc @anuchandy @ki1729

**Comment by zafar142007** [comment]
@joshfree Is there a timeline to fixing the issue that you mentioned? It is a critical issue for us and it would be helpful if you could give us a timeframe. @conniey Can you please mention which 'similar' issue is that?
cc @anuchandy @ki1729

**Comment by conniey** [comment]
The issue I was referring to is one fixed in #23332 and #23381. We have not released the azure-core-amqp package yet though.

**Comment by zafar142007** [comment]
@joshfree Could you please provide an ETA for this release of azure-core-amqp package? It is impacting us very badly and your timely help will be appreciated. cc @anuchandy @ki1729  @conniey

**Comment by anuchandy** [comment]
@zafar142007, just opened a pr addressing the thread leaks - https://github.com/Azure/azure-sdk-for-java/pull/23715. will update the thread once it is reviewed & merged; then, I'll share the nightly build if you have some cycles to validate it.

**Comment by zafar142007** [comment]
@anuchandy Sure, thanks. I will wait for the nightly build. cc  @ki1729  @conniey @joshfree

**Comment by zafar142007** [comment]
@anuchandy I see that PR https://github.com/Azure/azure-sdk-for-java/pull/23715 is merged, so I took the latest main from https://github.com/Azure/azure-sdk-for-java and built on my local to test

**Comment by ki1729** [comment]
@zafar142007 The release 7.4.0 was finished last Friday (https://search.maven.org/artifact/com.azure/azure-messaging-servicebus/7.4.0/jar). So we might be doing this as a patch release this week, will keep you posted on that, but meanwhile, I just want to make sure that the issue is mitigated on your end for now, using the nightly build?

**Comment by zafar142007** [comment]
@ki1729 Yes the thread leaks have ended when I tested with the nightly build. But can you check if there is still some memory leak as shown in the picture I posted?

**Comment by zafar142007** [comment]
@ki1729 Have the above changes been released?

**Comment by zafar142007** [comment]
@joshfree @anuchandy @conniey @ki1729 Can you please inform if these changes have been released?

**Comment by conniey** [comment]
Hey @zafar142007, our September release is slated for this coming week.

**Comment by ramya-rao-a** [comment]
Based on the [changelog](https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/servicebus/azure-messaging-servicebus/CHANGELOG.md#741-2021-09-13), looks like the fix for this was released as part of version 7.4.1 of the package
